cmd/kubeadm/app/util/config/cluster.go

```diff
@@ -180,23 +183,69 @@ func getNodeNameFromKubeletConfig(kubeconfigDir string) (string, error) {
 	return strings.TrimPrefix(cert.Subject.CommonName, constants.NodesUserPrefix), nil
 }
 
-// getAPIEndpoint returns the APIEndpoint for the current node
-func getAPIEndpoint(data map[string]string, nodeName string, apiEndpoint *kubeadmapi.APIEndpoint) error {
-	// gets the ClusterStatus from kubeadm-config
-	clusterStatus, err := UnmarshalClusterStatus(data)
+func getAPIEndpoint(client clientset.Interface, nodeName string, apiEndpoint *kubeadmapi.APIEndpoint) error {
+	var errs []error
+	if err := getAPIEndpointFromPodAnnotationWithRetry(client, nodeName, apiEndpoint); err == nil {
+		return nil
+	} else {
+		errs = append(errs, errors.WithMessagef(err, "could not retrieve API endpoints for node %q using pod annotations", nodeName))
+	}
+	if err := getAPIEndpointFromClusterStatus(client, nodeName, apiEndpoint); err == nil {
```

Comment:
I added a note as well (not yet pushed), but in any case, the most visible TODO's will cascade the rest of the removals because everything will fall apart when removing the annotated ones :)